### PR TITLE
[WIP] Update configuration change behavior

### LINKED ISSUE #2869: Improve how config changes are handled
Currently, when `cypress.json` file is changed, we re-open the project and close the browser if it is currently open and running a spec. We do this because a config option like `modifyObstructiveCode` or `chromeWebSecurity` could have changed, which requires restarting the project and/or browser. The downside is that the currently running spec is 'lost' and the user has to manually re-select it to run their tests again.

When the plugins file is changed, we reload the plugins file and it automatically takes effect without re-opening the project or closing the browser. This is arguably wrong because the user could have changed options that affect the project or the browser. New events coming in 4.0 (`run:start`, `spec:start`, etc) will also necessitate the need to 'restart' the run (re-open the project). It's also an issue because it's not clear to the user that anything has happened when they change their plugins file, so they might think they need to completely quit and re-open Cypress for their changes to take effect.

We should make this consistent and improve the user experience for both these cases by doing the following:

* On cypress.json/plugins file updates, notify user in the browser that changes have occurred, noting where they occurred (cypress.json, plugins file, etc). These changes will not be loaded yet.
* In that notification, give the user a button with the option to restart the browser and project. It will re-open the browser with the current spec.

### PR DESCRIPTION
**Pull request by lilaconlee**
- Closes #2869 

### User facing changelog

There is now a prompt to restart Cypress whenever changes are made to the project's configuration or plugin file instead of closing the Test Runner after changes. 

#### Problem

- Whenever changes are made to `cypress.json` currently, we close the Test Runner completely. This is abrupt and oftentimes unexpected. It actually just looks like Cypress crashed.
- Conversely, when changes are made to the `plugins/index.js` file, we just reload the file and *do not* close Cypress. This may mean that changes made to the plugins that only take effect when you restart the browser won't take effect and would likely cause confusion to the user. 

#### Solution

- Whenever changes are made to either the `cypress.json` or `plugins/index.js` file, we show a message in the Test Runner prompting them of the change and encouraging them to restart Cypress with a button.

#### Before when editing `cypress.json` or custom `configFile` in browser

- **Cypress kills Test Runner**

#### After when editing `cypress.json` or custom `configFile` in browser

Screen Shot 2019-12-31 at 3 37 47 PM

#### Before when editing `plugins/index.js` in browser

No notice, but since I edited the browser flags, the browser flags will actually take no effect and I will be confused.

Screen Shot 2019-12-31 at 3 40 42 PM

#### After when editing `plugins/index.js` in browser

Screen Shot 2019-12-31 at 3 40 15 PM

#### Before when editing `cypress.json` or custom `configFile` in Desktop-GUI

- DesktopGui refreshes with new value, possibly having an error shown

#### After when editing `cypress.json`, custom `configFile`, or plugins file in Desktop-GUI

Settings Tab

Screen Shot 2020-01-02 at 12 02 18 PM

Project Tab

Screen Shot 2020-01-02 at 12 06 19 PM

When multiple warnings, they just stack on top of each other.

Screen Shot 2020-01-06 at 4 36 32 PM

### PR Tasks

- [ ] Have tests been added/updated?
- [ ] Has the original issue been tagged with a release in ZenHub? 
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [ ] Have API changes been updated in the [`type definitions`](cli/types/index.d.ts)?
- [ ] Have new configuration options been added to the [`cypress.schema.json`](cli/schema/cypress.schema.json)?

### CODE REVIEW

File: packages/server/lib/gui/events.coffee [hunk 1]
```diff
@@ -54,6 +54,9 @@ handleEvent = (options, bus, event, id, type, arg) ->
     when "on:config:changed"
       onBus("config:changed")
 
+    when "reload:configuration:requested"
+      onBus("reload:configuration:requested")
```
Comment (brian-mann) [inline packages/server/lib/gui/events.coffee]: I see where you're going with this - you're going to emit this event to the server, which will then immediately send it back to you (called the same thing).

It's a bit indirect but perhaps it's necessarily for reasons I don't see. I was thinking that you don't actually emit this event from the desktop-gui - instead you just call into the handler that you have set here: https://github.com/cypress-io/cypress/pull/3211/files/cd6a8e7f12e3e0cdeea5a8de4a1d40f44593f01b..b9e98f8f5a3f8eea5296561d96c3ec302c5c59a7#diff-63b2e970dc3b8d943b276fa3766e32adR183

to handle that, you just need to extract out the function into another function that's called when you click the "Reload" button in the desktop GUI or when the `ipc` event comes back to you

**Comment by jennifer-shehane** [comment]
This needs revisiting - removed as needing review at this time.

**Comment by jennifer-shehane** [comment]
I updated the original comment to fill in the requirements of the pull request template.

Maybe I messed up some things upon merge conflicts, but some initial observations:

1. The path to the file shows the absolute path as opposed to the relative path to the project. This is not ideal. 
2. You can get rid of the notification by just refreshing / reloading the browser (without actually reloading the Cypress Test Runner). This does not seem ideal as it actually is not taking the changes into effect but the notification disappears. 
3. Clicking the restart button actually appears to do nothing (I probably broke this in merge conflicts). 
4. The notification does not show if a change was made to a `pluginsFile` other than the default one. So, I set a `pluginsFile` to a different path in my `cypress.json`, made a change to this file and there is no notification anymore. 
5. The notification actually does work on changes to a custom configuration file.

**Comment by jennifer-shehane** [comment]
I rewrote the reporter tests for the banner in Cypress and have confirmed that the `reload:configuration` event is being called when you click 'Restart', but the banner continues showing and does not go away. There never was a test for this, so this is likely a superficial UI problem rather than the config never reloading.

**Comment by jennifer-shehane** [comment]
Unfortunately we have to close this PR due to inactivity. Please open a new PR addressing the original issue and any requested changes.

**Pull request closed without merging**